README.org
#+TITLE: Omni-Quotes

[[https://travis-ci.org/AdrieanKhisbe/omni-quotes.el][file:https://travis-ci.org/AdrieanKhisbe/omni-quotes.el.svg]]
[[https://coveralls.io/r/AdrieanKhisbe/omni-quotes.el][file:https://coveralls.io/repos/AdrieanKhisbe/omni-quotes.el/badge.svg]]
[[http://melpa.org/#/omni-quotes][file:http://melpa.org/packages/omni-quotes-badge.svg]]
[[http://stable.melpa.org/#/omni-quotes][file:http://stable.melpa.org/packages/omni-quotes-badge.svg]]
[[https://github.com/AdrieanKhisbe/omni-quotes.el/tags][file:https://img.shields.io/github/tag/AdrieanKhisbe/omni-quotes.el.svg]]
[[http://www.gnu.org/licenses/gpl-3.0.html][http://img.shields.io/:license-gpl3-blue.svg]]

/Display all the Quotes!/

* Description

Omni Quotes is there to display quotes on you /Echo Area/ when you are idle.
Whether it's funny stuff, inspirational quotes, or something you try to remember.

* Usage

Just turn on =omni-quotes-mode=, and quotes will appear when you are idle enough.

You can load a quote file with the command =omni-quotes-load-simple-quote-file= and
reload default quotes with =omni-quotes-load-defaults=.
A quote file is just a basic text file with one quote per line.

When you have multiple quote-set it's possible to cycle through them with the commands
=omni-quotes-next-set= and =omni-quotes-prev-set=.


* Installation
You can install it from melpa, with =(package-install 'omni-quotes)=

If you use [[https://github.com/jwiegley/use-package][use-package]] you could use the following config
#+begin_src emacs-lisp
  (use-package omni-quotes
    :bind (("M-s q m" . omni-quotes-mode)
           ("M-s q p" . omni-quotes-prev-set)
           ("M-s q n" . omni-quotes-next-set)
           ("M-s q s" . omni-quotes-shuffle-current-set)
           ("M-s q q" . omni-quotes-display-random-quote))
    :config (omni-quotes-load-simple-quote-file "~/.my-quote" "personal"))
#+end_src

* Configuration
A set of custom is present to customize behavior and appearance of omniquotes.
You can access it through =customize-group 'omni-quotes= where you'll see the list of the customs.

Here are the main ones:

- =omni-quotes-default-quotes= :: List of default quotes
- =omni-quotes-prompt= :: Leading prompt of quotes messages
- =omni-quotes-idle-interval= :: Idle time after which omni-quote start to display quotes
- =omni-quotes-max-repeat= :: Number of omni-quotes will repeat without any activity. If zero there wont be any limit
- =omni-quotes-fading= :: Does the quote fade after some time
- =omni-quotes-fading-delay= :: Delay before the fading effect
- =omni-quotes-fading-duration= :: Time it takes for the quote to fade away
- =omni-quotes-repeat-interval= :: Time interval between two quotes
- =omni-quotes-boring-message-patterns= :: List of message that ca be overwritten by an omni-quote
- =omni-quotes-lighter= :: OmniQuote lighter (name in modeline).

If you want to customize them programaticaly in your config you must use the =customize-set-variable= construct, eg: =(customize-set-variable 'omni-quotes-fading t)=

* Notes

This is an inbuilding package, hence that huge number of building comments accross the code.
(usage of =omni-tags= is recommended :D).

*If you have any remark, refactor suggestion or else, just pose an issue ;)*


CHANGELOG.md
# Change Log

All notable changes to *omni-quotes.el* will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased][unreleased]
_Nothing so far_

## [0.5.1] - 2020-03-04
### Fixed
- Restored support to emacs26
- Avoid colision with other packages like ivy/swiper

## [0.5.0] - 2017-04-25
### Added
- Shuffle Quote set methods
- omni-quote-set-prev function
### Changed
- new implementation of shuffle list
### Fixed
- real iffle delay using post-command-hook
## [0.4.2] - 2017-04-20
### Added
- centered options
### CHANGED
- factorized customize setter
## [0.4.1] - 2017-04-19
### Fixed
- Omni-log fading minor bugs

## [0.4.0] - 2017-04-18
### Added
- Omni-log fading options

## [0.3.0] - 2017-04-16
### Added
- global quotes ring with possibility to go to next or previous set.
### Changed
- rename quote ring to quote set

## [0.2.0] - 2017-04-16
### Added
- feature to load quotes from a file
### Changed
- Introduce a quote-ring structure made with an hash table

## [0.1.0] - 2015-11-15
- "Initial" stable Release

[unreleased]: https://github.com/AdrieanKhisbe/omni-quotes.el/compare/v0.5.1...HEAD
[0.5.1]: https://github.com/AdrieanKhisbe/omni-quotes.el/compare/v0.5.0....v0.5.1
[0.5.0]: https://github.com/AdrieanKhisbe/omni-quotes.el/compare/v0.4.2....v0.5.0
[0.4.2]: https://github.com/AdrieanKhisbe/omni-quotes.el/compare/v0.4.1....v0.4.2
[0.4.1]: https://github.com/AdrieanKhisbe/omni-quotes.el/compare/v0.4.0....v0.4.1
[0.4.0]: https://github.com/AdrieanKhisbe/omni-quotes.el/compare/v0.3.0....v0.4.0
[0.3.0]: https://github.com/AdrieanKhisbe/omni-quotes.el/compare/v0.2.0....v0.3.0
[0.2.0]: https://github.com/AdrieanKhisbe/omni-quotes.el/compare/v0.1.1....v0.2.0
[0.1.0]: https://github.com/AdrieanKhisbe/omni-quotes.el/compare/7ca58b7....v0.1.0


test/test-helper.el
;; test-helper - for omni-quotes

;;; Code:

(require 'f)

(defvar omni-quotes-path
  (f-parent (f-this-file)))

(defvar omni-quotes-test-path
  (f-dirname (f-this-file)))

(defvar omni-quotes-root-path
  (f-parent omni-quotes-test-path))

(require 'undercover)
(undercover "*.el" "omni-quotes/*.el"
            (:exclude "*-test.el")
            (:send-report nil)
            (:report-file "/tmp/undercover-report.json"))


(require 'ert)
(require 'cl)
(require 's)
(require 'dash)
(require 'shut-up)
(require 'el-mock)
(require 'omni-log)
(require 'omni-quotes-ring (f-expand "omni-quotes-ring" omni-quotes-root-path))
(require 'omni-quotes-timer (f-expand "omni-quotes-timer" omni-quotes-root-path))
(require 'omni-quotes-reader (f-expand "omni-quotes-reader" omni-quotes-root-path))
(require 'omni-quotes (f-expand "omni-quotes" omni-quotes-root-path))

;;; test-helper.el ends here


test/omni-quotes-test.el
(ert-deftest  can-retrieve-quote ()
  (should (s-presence (omni-quotes-random-quote))))

(ert-deftest can-set-custom ()
  (customize-set-variable 'omni-quotes-fading t)
  (should (equal t omni-quotes-fading)))

;;; ¤> parsers

(ert-deftest load-simple-quote-file ()
  (mocklet (((omni-quotes-set-populate '("a" "b" "c" "alpha" "beta") "quote-file")))
           (omni-quotes-load-simple-quote-file (f-expand "quote-file.txt" omni-quotes-test-path) "quote-file")))

(ert-deftest can-parse-simple-file-ok ()
  (should (equal (omni-quotes-simple-parser (f-expand "quote-file.txt" omni-quotes-test-path))
                 '("a" "b" "c" "alpha" "beta"))))

(ert-deftest can-parse-simple-file-ko ()
  (should-not (shut-up (omni-quotes-simple-parser (f-expand "NOT-A-quote-file.txt" omni-quotes-test-path)))))


;;; ¤> quotes display

(ert-deftest omni-quotes-cant-redisplay-no-message ()
  (with-mock
   (stub current-message => nil)
   (should-not (omni-quotes-cant-redisplay))))

(ert-deftest omni-quotes-cant-redisplay-boring-message ()
  (with-mock
   (stub current-message => "Mark set")
   (should-not (omni-quotes-cant-redisplay))))

(ert-deftest omni-quotes-cant-redisplay-important-message ()
  (with-mock
   (stub current-message => "Important Message")
   (should (omni-quotes-cant-redisplay))))

(ert-deftest omni-quote-idle-redisplay-active-minibuffer()
  (mocklet ((active-minibuffer-window => t)
            (omni-quotes-display-random-quote not-called))
           (omni-quotes-idle-display-callback)))

(ert-deftest omni-quote-idle-redisplay-cant()
  (mocklet ((active-minibuffer-window => nil)
            (omni-quotes-cant-redisplay => t)
            (omni-quotes-display-random-quote not-called))
           (omni-quotes-idle-display-callback)))

(ert-deftest omni-quote-idle-redisplay-cant()
  (mocklet ((active-minibuffer-window => nil)
            (omni-quotes-cant-redisplay => nil)
            (omni-quotes-display-random-quote => "called"))
           (should (equal (omni-quotes-idle-display-callback) "called"))))

(ert-deftest calling-omni-quote-start-cancels-the-timer()
  (omni-quotes-idle-display-start)
  (should omni-quotes-idle-timer)
  (omni-quotes-idle-display-start)
  (should omni-quotes-idle-timer))

;;; ¤> Quote ring

(ert-deftest omni-quotes-next-set-ok ()
  (let ((pointer omni-quotes-sets-ring-pointer))
    (omni-quotes-next-set)
    (should (equal (1+ pointer) omni-quotes-sets-ring-pointer))))

(ert-deftest omni-quotes-prev-set-ok ()
  (let ((pointer omni-quotes-sets-ring-pointer))
    (omni-quotes-prev-set)
    (should (equal (1- pointer) omni-quotes-sets-ring-pointer))))

;; ¤>> simple test, calling function
(ert-deftest shuffle-didnt-fail ()
  (omni-quotes-shuffle-current-set)
  (should (equal (ht-get omni-quotes-current-set 'pointer) 0)))

(ert-deftest omni-quote-set-prev-test()
  (let ((pointer (ht-get omni-quotes-current-set 'pointer)))
    (should (omni-quotes-set-prev omni-quotes-current-set))
    (should (equal (1- pointer) (ht-get omni-quotes-current-set 'pointer)))))

(ert-deftest omni-quote-set-random-test()
    (should (omni-quotes-set-random omni-quotes-current-set)))


myquote.el
;;; §draft. idee de quote à charger.
;;; display random au fur et à mesure. ? ring/list
;;; Utiliser aussi pour commandes à renforcer
;;; trigger. affiche après un certain random.
;;; §Later: plusieurs catégories.
;;; §maybe: le binder avec un des trucs de quotes?
;;; use dash library

(defvar my-quotes nil "My stupid quotes")
(setq my-quotes '(
		  ;; Emacs custos
		  "Customization is the corner stone of Emacs"
		  "Emacs is an acronym for \"Escape Meta Alt Control Shift\""

		  ;; Tips
		  "Harness Macro Powaaaa"
		  "Use a fuckying good register level" ; paye ton franglish
		  "Might be to learn to make function from macros"
		  "Aren't you a bit déRanger?...(c'est le bo'del)"

		  ))

(defun my-random-quote ()
  "Affiche une quote au hasard"
  (interactive)
  (message (nth (random (length my-quotes)) my-quotes)))
;; see berkeley: utilities.lisp!!!

;; Following timer recipe: (§todo:maybe extract a library?)

;; variable for the timer object
(defvar idle-timer-quote-timer nil "MyQuote timer")
(defvar quote-idle-time 5 "MyQuote idle time, in seconds")
(defvar quote-repeat-time 30 "MyQuote repeat time, in seconds")
;; §todo: to custim

(defun idle-timer-quote-callback ()
  "MyQuote Timer Call bac function"
  ;; §TODO! check if there is no prompt waiting!!
  (my-random-quote))

;; §keep?
(defun idle-timer-quote-run-once ()
  (interactive)
  (cancel-if-quote-timer)
  ;; maybe extract update timer function?
  (setq idle-timer-quote-timer
	(run-with-idle-timer quote-idle-time nil #'idle-timer-quote-callback)))

(defun idle-timer-quote-start ()
  (interactive)
  (cancel-if-quote-timer)
  (setq idle-timer-quote-timer
	(run-with-timer quote-idle-time quote-repeat-time #'idle-timer-quote-callback)))

;; stop function
(defun idle-timer-quote-stop ()
  (interactive)
  (cancel-if-quote-timer)
  (setq idle-timer-quote-timer nil))

;; utils
(defun cancel-if-quote-timer ()
  (when (timerp idle-timer-quote-timer)
    (cancel-timer idle-timer-quote-timer)))

;; §Todo: minor mode.

(provide 'myquote)


omni-quotes-ring.el
;;; omni-quotes-ring.el --- Datastructure to old [Omni] Quotes
;;
;; Copyright (C) 2014-2020 Adrien Becchis
;;
;; Author: Adrien Becchis <adriean.khisbe@live.fr>
;; Keywords: convenience

;; This file is not part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;;; Code:

(require 'dash)
(require 'ht)

;;; ¤>vars
(defvar omni-quotes-current-set nil
  "Quote Sets Storing the Different quotes.")

(defvar omni-quotes-sets (ht)
  "Hashtable of the quote sets.")

(defvar omni-quotes-sets-ring (make-ring 42)
  "Ring of the used quote sets.")

(defvar omni-quotes-sets-ring-pointer 0
  "Pointer on the `omni-quote-sets-ring'")

(defun omni-quotes-set-populate (quote-list name)
  "Populate `omni-quotes-current-set' with a quote set made out of
the provided QUOTE-LIST and NAME."
  (let ((quote-set (omni-quote-set-maker quote-list name)))
    ;; §todo: protect from nil
    (ht-set! omni-quotes-sets name quote-set)
    (ring-insert omni-quotes-sets-ring quote-set)
    (setq omni-quotes-current-set quote-set)))

(defun omni-quote-set-maker (list name)
  "Make a Quote-Set out of the provided LIST."
  (let ((ring (ht ('list list)
                  ('name name)
                  ('pointer 0)
                  ('ring (make-ring (length list))))))
    (-each (omni-quote--shuffle-list list)
      (lambda(quote)(ring-insert (ht-get ring 'ring) quote)))
    ring))

(defun omni-quote--shuffle-list (list)
  "Returns a shuffled version of the LIST."
  (let ((new-list (reverse list))) ; simple clone (suffling is in place)
    (-each (reverse (number-sequence 1 (1- (length new-list))))
      (lambda (i) "Swap two elements."
        (let* ((j (random (+ i 1)))
               (tmp (elt new-list i)))
          (setf (elt new-list i) (elt new-list j))
          (setf (elt new-list j) tmp))))
    new-list))


(defun omni-quotes-set-next (quote-set)
  "Send current quote of the QUOTE-SET and move pointer forward."
  (let* ((ring (ht-get quote-set 'ring))
         (pointer (ht-get quote-set 'pointer))
         (quote (ring-ref ring pointer)))
    (ht-set! quote-set 'pointer (1+ pointer))
    quote))

(defun omni-quotes-set-prev (quote-set)
  "Send current quote of the QUOTE-SET and move pointer backward."
  (let* ((ring (ht-get quote-set 'ring))
         (pointer (ht-get quote-set 'pointer))
         (quote (ring-ref ring pointer)))
    (ht-set! quote-set 'pointer (1- pointer))
    quote))

(defun omni-quotes-set-random (quote-set)
  "Give a random quote from the QUOTE-SET."
  (let ((ring (ht-get quote-set 'ring)))
    (ring-ref ring (random (ring-size ring)))))

(defun omni-quotes-set-get (quote-set)
  "Get a quote from the given QUOTE-SET."
  ;; §maybe: different accès method. Get method dispatch
  (omni-quotes-set-next quote-set)

  ;; §later: var saying method that should be call
  )
;; ¤see: berkeley: utilities.lisp!!!

(defun omni-quotes-shuffle-set (quote-set)
  "Reshuffle given QUOTE-SET."
  (let* ((new-list (omni-quote--shuffle-list (ht-get quote-set 'list)))
         (new-ring (make-ring (length new-list))))
    (-each new-list (lambda (quote) (ring-insert new-ring quote)))
    (ht-set! quote-set 'ring new-ring)
    (ht-set! quote-set 'pointer 0)))

(defun omni-quotes-shuffle-current-set ()
  "Reshuffle current `omni-quotes-current-set'."
  (interactive)
  (omni-quotes-shuffle-set omni-quotes-current-set))

(defun omni-quotes-next-set ()
  "Shift the `omni-quotes-current-set' forward."
  (interactive) ; §todo: universal arg.
  (let ((new-pointer (1+ omni-quotes-sets-ring-pointer)))
    (setq omni-quotes-sets-ring-pointer new-pointer
          omni-quotes-current-set (ring-ref omni-quotes-sets-ring new-pointer))))

(defun omni-quotes-prev-set ()
  "Shift the `omni-quotes-current-set' backward."
  (interactive)
  (let ((new-pointer (1- omni-quotes-sets-ring-pointer)))
    (setq omni-quotes-sets-ring-pointer new-pointer
          omni-quotes-current-set (ring-ref omni-quotes-sets-ring new-pointer))))

(provide 'omni-quotes-ring)
;;; omni-quotes-ring.el ends here


omni-quote-reader.el
;;; omni-quotes-reader.el --- Utilities to read "quotes" files
;;
;; Copyright (C) 2014 Adrien Becchis
;;
;; Author: Adrien Becchis <adriean.khisbe@live.fr>
;; Keywords: convenience

;; This file is not part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;; This file is used to read quotes from differents kind of files formats
;; some custom and some not.

;;; Code:

;; §later: see existing format.


;; see how

;; ¤> simple file reader:
;; format is a line a quote.
;; ignore file w


;; ¤parser function should send back
(defun oq:process-quote-file (file parser)
  ;; ¤doc!
  ;; §todo: type checking
  (with-temp-buffer
    (insert-file-contents file)
    (goto-char (point-min))
    (funcall parser))

  ;; §see: what do send back: result. or stored already?
  )

(defun oq:parser:a-quote-aline ()
  ;; see comment traiter ligne par ligne?
  ;; ou tuiliser regexp
  (let ((quotes '()) )
    (while (not (eobp))
      ;; (buffer-substring)
      (push (thing-at-point 'line) quotes)
      ;; §later: filter
      (forward-line))
    quotes))


(defun oq:test ()
  (oq:process-quote-file
   "~/AA"
   'oq:parser:a-quote-aline
   ))
;; §todo: create tests
;; §how?
(oq:test)
;; §pb: grab the end of line. should delete this.


(provide 'omni-quotes-reader)
;;; omni-quotes-reader.el ends here


omni-quotes-timer.el
;;; omni-quotes-timer.el --- Timer functions for OmniQuotes
;;
;; Copyright (C) 2014-2020 Adrien Becchis
;;
;; Author: Adrien Becchis <adriean.khisbe@live.fr>
;; Keywords: convenience

;; This file is not part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;; There are the timer code related of omni-quotes

;;; Code:


(defvar omni-quotes-idle-timer nil "OmniQuote timer.")

(defun  omni-quotes-idle-display-start (&optional no-repeat)
  "Add OmniQuote idle timer with repeat (by default).

With NO-REPEAT idle display will happen once."
  (when (timerp omni-quotes-idle-timer)
    (cancel-timer omni-quotes-idle-timer))
  (setq omni-quotes-idle-timer (run-with-timer omni-quotes-idle-interval
                                               (if no-repeat nil omni-quotes-repeat-interval)
                                               #'omni-quotes-idle-display-callback)))

(defun omni-quotes-idle-display-stop ()
  "Stop OmniQuote Idle timer."
  (when (timerp omni-quotes-idle-timer)
    (cancel-timer omni-quotes-idle-timer))
  (setq omni-quotes-idle-timer nil))

(provide 'omni-quotes-timer)
;;; omni-quotes-timer.el ends here


omni-quotes-reader.el
;;; omni-quotes-reader.el --- Utilities to read "quotes" files
;;
;; Copyright (C) 2014-2020 Adrien Becchis
;;
;; Author: Adrien Becchis <adriean.khisbe@live.fr>
;; Keywords: convenience

;; This file is not part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;; This file is used to read quotes from differents kind of files formats
;; some custom and some not.

;;; Code:

(require 'f)
(require 'dash)

;; §later: see existing format of quotes

;;; ¤> Entry points.
(defun omni-quotes-load-simple-quote-file (file-name name)
  "Loads quotes from given FILE-NAME as current quote-ring"
  (interactive "fQuote File: \nsQuoteSet name: ")
  (let ((quotes-list (omni-quotes-simple-parser file-name)))
    (omni-quotes-set-populate quotes-list name)))

(defun omni-quotes-load-defaults ()
  "Loads the defaults quote as current quote-set."
  (interactive)
  (omni-quotes-set-populate omni-quotes-default-quotes "default"))

;;; ¤> parsers
(defun omni-quotes-simple-parser (file-name)
  "Returns a list of quote from a simple FILE-NAME."
  (if (f-exists? file-name)
      (let ((text (f-read-text file-name)))
        (s-lines (s-trim text)))
    ;; §todo: filter pattern
    (progn
      (message "Filename does not exists %s" file-name)
      nil)))

(provide 'omni-quotes-reader)
;;; omni-quotes-reader.el ends here


omni-quotes.el
;;; omni-quotes.el --- Random quotes displayer   -*- lexical-binding: t -*-
;;
;; Copyright (C) 2014-2020 Adrien Becchis
;;
;; Author: Adrien Becchis <adriean.khisbe@live.fr>
;; Created: 2014-07-17
;; Last-Release: 2020-03-04
;; Version: 0.5.1
;; Keywords: convenience
;; Package-Requires: ((dash "2.8") (omni-log "0.4.0") (f "0.19.0") (s "1.11.0") (ht "2.1"))
;; Url: https://github.com/AdrieanKhisbe/omni-quotes.el


;; This file is not part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Building-Notes:
;; §Later: plusieurs catégories.
;; §maybe: le binder avec un des trucs de quotes?: fortunes, and co.

;;; Commentary:
;; Omni Quotes is there to display quotes on your Echo Area when emacs is idle,
;; whether it's funny stuff, inspirational quotes, or something you try to remember.

;;; Code:

(require 'dash)
(require 's)
(require 'f)
(require 'omni-log)
(require 'omni-quotes-timer)
(require 'omni-quotes-ring)
(require 'omni-quotes-reader)

;;; ¤> customs:
(defcustom omni-quotes-lighter " Ξ" "OmniQuote lighter (name in modeline) if any."
  ;; §maybe: replace by Q, (or greek style ϙ)
  :type 'string :group 'omni-quotes)

(defcustom omni-quotes-idle-interval 4 "OmniQuote idle time, in seconds."
  :type 'number :group 'omni-quotes)

(defcustom omni-quotes-repeat-interval 20 "OmniQuote repeat time, in seconds."
  :type 'number :group 'omni-quotes)

(defun omni-quotes--log-setter (property)
  "Return function handling forwarding PROPERTY update to logger."
  (lambda (symb value)
    (if (boundp 'omni-quotes-global-quote-log)
        (omni-log-logger-set-property omni-quotes-global-quote-log property value))
    (set-default symb value)))

(defcustom omni-quotes-prompt " » " "Leading prompt of OmniQuotes messages."
  :type 'string :group 'omni-quotes
  :set (lambda (symb value)
         (let ((value (propertize value 'omni-quote-p t)))
           (if (boundp 'omni-quotes-global-quote-log)
               (omni-log-logger-set-property omni-quotes-global-quote-log 'prompt value))
           (set-default symb value))))

(defcustom omni-quotes-max-repeat 6
  "Number of omni-quotes will repeat without any activity. If zero there wont be any limit"
  :type 'number :group 'omni-quotes)

(defcustom omni-quotes-fading nil
  "Does omni-quote fade after some duration."
  :type 'boolean :group 'omni-quotes
  :set (omni-quotes--log-setter 'fading))

(defcustom omni-quotes-fading-delay 14
  "Delay after which quote will fade away."
  :type 'number :group 'omni-quotes
  :set (omni-quotes--log-setter 'fading-delay))

(defcustom omni-quotes-fading-duration 4
  "Duration of the fade away effect."
  :type 'number :group 'omni-quotes
  :set (omni-quotes--log-setter 'fading-duration))

(defcustom omni-quotes-centered nil
  "Does omni-quote fade after some duration."
  :type 'boolean :group 'omni-quotes
  :set (omni-quotes--log-setter 'centered))

(defcustom omni-quotes-default-quotes
  '(
    ;; Emacs custos
    "Customization is the corner stone of Emacs"
    "Emacs is an acronym for \"Escape Meta Alt Control Shift\""

    ;; Tips
    "Harness Macro Powaaaa"
    "Register registers as a good practice"
    "Might be to learn to make function from macros"

    "Use position register and jump everywhere in no time! :)"
    "Bookmark are a must learn feature!"
    "Repeat command with C-x zzzzzzzzzzzzzzzzzzzzzz. (and don't fall asleep)"

    "Get some projectile and don't throw them away!"
    "Ace and jump chars!"
    "Go to the Helm Gouffre!" ; ref to lord of rings
    "Don't be Evil (nor a God)"
    ) ; end-of default quotes
  "Some stupid default (omni-)quotes."
  :type '(repeat string) :group 'omni-quotes)

;; §later: custom: quotes sources
;; §later: custom whitelist messages to bypass!!

;; §later: use category. (revision, stupid quote, emacs tips, emacs binding to learn...)
;;         category based on context (ex langage specific tips)

(defcustom omni-quotes-boring-message-patterns
  '(
    "^Omni-Quotes mode enabled"
    "^Mark set"
    "^Auto-saving...done"
    "^Quit"
    "End of buffer"
    "^For information about GNU Emacs"
    ;; yas
    "^\\[yas\\]"
    ;; use-package
    "^Configuring package" "^Loading package" "^use-package idle:"
    "^Here is not Git/Mercurial work tree"
    "^Saving file" "^Wrote /"
    )
  "List of message that can be overwrite by an OmniQuote."
  :type '(repeat regexp) :group 'omni-quotes
  )

(defvar omni-quotes-boring-message-regexp
  (mapconcat 'identity omni-quotes-boring-message-patterns  "\\|")
  "Regexp used to match messages that can be overwriten by a quote.
Constructed from `omni-quotes-boring-message-patterns'.")

(defconst omni-quotes-global-quote-log
  (omni-log-create "omni-quotes"
                   `((prompt . ,omni-quotes-prompt)
                     (fading . ,omni-quotes-fading)
                     (fading-delay . ,omni-quotes-fading-delay)
                     (fading-duration . ,omni-quotes-fading-duration)
                     (centered . ,omni-quotes-centered)))
      "Specific logger for omni-quotes.")

;;;###autoload
(defun omni-quotes-display-random-quote ()
  "Display a random quote obtained from `omni-quotes-random-quote'.
The quote will be prefixed by the current `omni-quotes-prompt'"
  (interactive)
  (log-omni-quotes (omni-quotes-random-quote)))
;; §maybe: [append with date?]
;; §maybe: change format: catégorie > texte.

(defun omni-quotes-random-quote ()
  "Get a random quote."
  (omni-quotes-set-get omni-quotes-current-set))
;; §maybe: should have different quote rings for the categories. how to select-active-regions
;; §maybe have current function: (round, random...)
;; §maybe: create an intensive mode. quotes plus raprochées. éventuellement un slidding effect. sans interruption
;;        jusqu'à la prochaine touche

(defvar omni-quotes--nb-current-repeat 0)

(defun omni-quotes-idle-display-callback () ; §maybe rename of move in timer?
  "OmniQuote Timer callback function."
  ;; §maybe: force? optional argument? §maybe: extract in other function

  (if (or (active-minibuffer-window) ; check if there is no prompt waiting
              (omni-quotes-cant-redisplay))
      (setq omni-quotes--nb-current-repeat 0)
    (let ((cm (current-message)))
      (setq omni-quotes--nb-current-repeat
            (if (and cm (get-text-property 0 'omni-quote-p cm)) (1+ omni-quotes--nb-current-repeat) 1))
      (if (or (eq 0 omni-quotes-max-repeat)
              (>= omni-quotes-max-repeat omni-quotes--nb-current-repeat))
          (omni-quotes-display-random-quote)))))

(defun omni-quotes-cant-redisplay()
  "Tells if Quote should be display. (in order to avoid erasing of important messages)"
  (let ((cm (current-message)))
    (and cm
         (not (or (get-text-property 0 'omni-quote-p cm)
                  (string-match omni-quotes-boring-message-regexp cm))))))


;;;###autoload
(define-minor-mode omni-quotes-mode
  "Display random quotes when idle."
  :lighter omni-quotes-lighter
  :global t
  (if omni-quotes-mode
      (progn
        (add-hook 'post-command-hook 'omni-quotes-idle-display-start)
        (omni-quotes-idle-display-start))
    (progn
      (remove-hook 'post-command-hook 'omni-quotes-idle-display-start)
      (omni-quotes-idle-display-stop))))

;; §maybe: tmp?
(omni-quotes-load-defaults)

(provide 'omni-quotes)
;;; omni-quotes.el ends here
